hubblestack/extmods/grains/systemuuid.py

```diff
@@ -2,40 +2,85 @@
 '''
 Gather the system uuid via osquery
 '''
+import logging
+import os
 import salt.utils.path
 import salt.modules.cmdmod
 
 __salt__ = {'cmd.run': salt.modules.cmdmod._run_quiet}
+log = logging.getLogger(__name__)
 
 
 def get_system_uuid():
     '''
-    Gather the system uuid via osquery
+    Gather the system uuid via osquery and store it on disk.
 
     If osquery can't get a hardware-based value, it'll just randomly generate a new uuid every time.
     If that happens, fall back to the hubble_uuid.
     '''
     # Provides:
     #   system_uuid
 
-    options = '"SELECT uuid AS system_uuid FROM osquery_info;" --header=false --csv'
+    cached_system_uuid_path = os.path.join(os.path.dirname(__opts__['configfile']),
+                                           'hubble_cached_system_uuid')
+    previous_system_uuid = __opts__.get('system_uuid', None)
+
+    # Get the system uuid via osquery. If it changes, fall back to hubble_uuid
+    live_system_uuid = _get_uuid_from_system() or __opts__.get('hubble_uuid', None)
+
+    if not live_system_uuid:
```

Comment:
Does this mean that on hosts where getting uuid is broken there is no `system_uuid` grain for at least until the grains get refreshed the second time? 

Would a check comparing `hubble_cached_uuid` and `hubble_cached_system_uuid` files make sense?

Reply:
Yes, that's correct. But that's how we've always done things I think. We decided that we want system_uuid to equal hubble_uuid in the cases where system_uuid changes. If system_uuid is changing, and there is no hubble_uuid yet, we can't do that, so we just wait until next time.

Reply:
`hubble_uuid` may be waiting in `hubble_cached_system_uuid` file, if it got past the check at startup.

Reply:
I don't think it's a good idea to have the system_uuid grain reach in and open the hubble_uuid cache file.

Reply:
What is the problem with reading that cache file?

Reply:
I just think it's dirty to read the cache file for the `hubble_uuid` grain in the `system_uuid` grain code. But I did it anyway.